File: internal/ns/open_panel_darwin.go
// Copyright ©2021 by Richard A. Wilkes. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, version 2.0. If a copy of the MPL was not distributed with
// this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This Source Code Form is "Incompatible With Secondary Licenses", as
// defined by the Mozilla Public License, version 2.0.

package ns

import "github.com/progrium/macdriver/objc"

var openPanelClass = objc.Get("NSOpenPanel")

// OpenPanel https://developer.apple.com/documentation/appkit/nsopenpanel?language=objc
type OpenPanel struct {
	SavePanel
}

// NewOpenPanel https://developer.apple.com/documentation/appkit/nsopenpanel/1584365-openpanel?language=objc
func NewOpenPanel() OpenPanel {
	return OpenPanel{SavePanel: SavePanel{Object: openPanelClass.Send("openPanel")}}
}

// CanChooseFiles https://developer.apple.com/documentation/appkit/nsopenpanel/1527060-canchoosefiles?language=objc
func (p OpenPanel) CanChooseFiles() bool {
	return p.Send("canChooseFiles").Bool()
}

// SetCanChooseFiles https://developer.apple.com/documentation/appkit/nsopenpanel/1527060-canchoosefiles?language=objc
func (p OpenPanel) SetCanChooseFiles(set bool) {
	p.Send("setCanChooseFiles:", set)
}

// CanChooseDirectories https://developer.apple.com/documentation/appkit/nsopenpanel/1532668-canchoosedirectories?language=objc
func (p OpenPanel) CanChooseDirectories() bool {
	return p.Send("canChooseDirectories").Bool()
}

// SetCanChooseDirectories https://developer.apple.com/documentation/appkit/nsopenpanel/1532668-canchoosedirectories?language=objc
func (p OpenPanel) SetCanChooseDirectories(set bool) {
	p.Send("setCanChooseDirectories:", set)
}

// ResolvesAliases https://developer.apple.com/documentation/appkit/nsopenpanel/1533625-resolvesaliases?language=objc
func (p OpenPanel) ResolvesAliases() bool {
	return p.Send("resolvesAliases").Bool()
}

// SetResolvesAliases https://developer.apple.com/documentation/appkit/nsopenpanel/1533625-resolvesaliases?language=objc
func (p OpenPanel) SetResolvesAliases(set bool) {
	p.Send("setResolvesAliases:", set)
}

// AllowsMultipleSelection https://developer.apple.com/documentation/appkit/nsopenpanel/1530786-allowsmultipleselection?language=objc
func (p OpenPanel) AllowsMultipleSelection() bool {
	return p.Send("allowsMultipleSelection").Bool()
}

// SetAllowsMultipleSelection https://developer.apple.com/documentation/appkit/nsopenpanel/1530786-allowsmultipleselection?language=objc
func (p OpenPanel) SetAllowsMultipleSelection(set bool) {
	p.Send("setAllowsMultipleSelection:", set)
}

// URLs https://developer.apple.com/documentation/appkit/nsopenpanel/1529845-urls?language=objc
func (p OpenPanel) URLs() Array {
	return Array{Object: p.Send("URLs")}
}

File: internal/mac/open_panel_darwin.m
// Copyright (c) 2021-2026 by Richard A. Wilkes. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, version 2.0. If a copy of the MPL was not distributed with
// this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This Source Code Form is "Incompatible With Secondary Licenses", as
// defined by the Mozilla Public License, version 2.0.

#import "macos.h"

NSOpenPanelRef newOpenPanel() {
	return [[NSOpenPanel openPanel] retain];
}

CFURLRef openPanelDirectoryURL(NSOpenPanelRef openPanel) {
	return (CFURLRef)[(NSOpenPanel *)openPanel directoryURL];
}

void openPanelSetDirectoryURL(NSOpenPanelRef openPanel, CFURLRef url) {
	[(NSOpenPanel *)openPanel setDirectoryURL:(NSURL *)url];
}

CFArrayRef openPanelAllowedFileTypes(NSOpenPanelRef openPanel) {
	return (CFArrayRef)([(NSOpenPanel *)openPanel allowedFileTypes]);
}

void openPanelSetAllowedFileTypes(NSOpenPanelRef openPanel, CFArrayRef types) {
	[(NSOpenPanel *)openPanel setAllowedFileTypes:(NSArray<NSString *>*)(types)];
}

bool openPanelCanChooseFiles(NSOpenPanelRef openPanel) {
	return [(NSOpenPanel *)openPanel canChooseFiles];
}

void openPanelSetCanChooseFiles(NSOpenPanelRef openPanel, bool set) {
	[(NSOpenPanel *)openPanel setCanChooseFiles:set];
}

bool openPanelCanChooseDirectories(NSOpenPanelRef openPanel) {
	return [(NSOpenPanel *)openPanel canChooseDirectories];
}

void openPanelSetCanChooseDirectories(NSOpenPanelRef openPanel, bool set) {
	[(NSOpenPanel *)openPanel setCanChooseDirectories:set];
}

bool openPanelResolvesAliases(NSOpenPanelRef openPanel) {
	return [(NSOpenPanel *)openPanel resolvesAliases];
}

void openPanelSetResolvesAliases(NSOpenPanelRef openPanel, bool set) {
	[(NSOpenPanel *)openPanel setResolvesAliases:set];
}

bool openPanelAllowsMultipleSelection(NSOpenPanelRef openPanel) {
	return [(NSOpenPanel *)openPanel allowsMultipleSelection];
}

void openPanelSetAllowsMultipleSelection(NSOpenPanelRef openPanel, bool set) {
	[(NSOpenPanel *)openPanel setAllowsMultipleSelection:set];
}

CFArrayRef openPanelURLs(NSOpenPanelRef openPanel) {
	return (CFArrayRef)[(NSOpenPanel *)openPanel URLs];
}

bool openPanelRunModal(NSOpenPanelRef openPanel) {
	return [(NSOpenPanel *)openPanel runModal] == NSModalResponseOK;
}
